PurchasesCoreSwift/Caching/DeviceCache.swift

```diff
@@ -0,0 +1,491 @@
+//
+//  DeviceCache.swift
+//  PurchasesCoreSwift
+//
+//  Created by Joshua Liebowitz on 7/13/21.
+//  Copyright © 2021 Purchases. All rights reserved.
+//
+
+import Foundation
+
+// swiftlint:disable file_length
+@objc(RCDeviceCache) open class DeviceCache: NSObject {
+
+    // Thread-safe, but don't call from anywhere inside this class.
+    @objc open var cachedAppUserID: String? { readCache { self.threadUnsafeCachedAppUserID } }
+    @objc open var cachedLegacyAppUserID: String? {
+        return readCache {
+            return self.userDefaults.string(forKey: CacheKeys.legacyGeneratedAppUserDefaults)
+        }
+    }
+    @objc open var cachedOfferings: Offerings? { offeringsCachedObject.cachedInstance() }
+
+    private let cacheDurationInSecondsInForeground = 60 * 5.0
+    private let cacheDurationInSecondsInBackground = 60 * 60 * 25.0
+    private let accessQueue = DispatchQueue(label: "DeviceCacheQueue")
+    private var threadUnsafeCachedAppUserID: String? { userDefaults.string(forKey: CacheKeys.appUserDefaults.rawValue) }
+    private let userDefaults: UserDefaults
+    private let notificationCenter: NotificationCenter
+    private let offeringsCachedObject: InMemoryCachedObject<Offerings>
+    private var appUserIDHasBeenSet: Bool = false
+    private let assertionFunction: (String) -> Void
+
+    @objc convenience public init(userDefaults: UserDefaults = UserDefaults.standard) {
+        self.init(userDefaults: userDefaults, offeringsCachedObject: nil, notificationCenter: nil)
+    }
+
+    public init(userDefaults: UserDefaults = UserDefaults.standard,
+                offeringsCachedObject: InMemoryCachedObject<Offerings>? = InMemoryCachedObject(),
+                notificationCenter: NotificationCenter? = NotificationCenter.default,
+                assertionFunction: @escaping (String) -> Void = { assertionFailure($0) }) {
+
+        self.offeringsCachedObject = offeringsCachedObject ?? InMemoryCachedObject()
+        self.notificationCenter = notificationCenter ?? NotificationCenter.default
+        self.userDefaults = userDefaults
+        self.assertionFunction = assertionFunction
+        self.appUserIDHasBeenSet = userDefaults.string(forKey: CacheKeys.appUserDefaults) != nil
+
+        super.init()
+
+        self.notificationCenter.addObserver(self,
+                                            selector: #selector(handleUserDefaultsChanged),
+                                            name: UserDefaults.didChangeNotification,
+                                            object: self.userDefaults)
+    }
+
+    @objc private func handleUserDefaultsChanged(notification: Notification) {
+        guard let notificationObject = notification.object as? UserDefaults else {
+            return
+        }
+
+        if appUserIDHasBeenSet && notificationObject == self.userDefaults {
```

Comment:
nitpick: this bit might look a bit clearer as two distinct sets of conditions: 
1. it's the notification we care about
2. the app user id that we set has been removed

```swift
guard let notificationObject = notification.object as? UserDefaults, 
    notificationObject == self.userDefaults else { 
    return
}

if appUserIDHasBeenSet && threadUnsafeCachedAppUserID == nil {
    assertionFunction(...
}
```

Reply:
Ay, good catch.